# mk-ORegAnno_2\_bed - module for converting ORegAnno data base to VEP friendly format


## About mk-ORegAnno_2\_bed

This module converts the ORegAnno tab delimited database into a bed file that can be used for [VEP custom annotation](https://www.ensembl.org/info/docs/tools/vep/script/vep_custom.html) ([1])

[The Open Regulatory Annotation database (OregAnno)](http://www.oreganno.org/)([2]) ORegAnno is a crowd-curated regulatory annotation database. It contains information about regulatory regions, transcription factor binding sites, RNA binding sites, regulatory variants, haplotypes, and other regulatory elements. Coordinates are relative to the GRCh37 assembly. However the original database contains information for other species and for assembly hg18 and GRCh38. The database implemented is the 2015.19.15.tsv version, downloaded from <http://www.oreganno.org/dump/>.

Genome version in ORegAnno depends on species and genome build, as for May 16th 2018, ORegAnno includes information for the following genomes:

**Species Build (version)**

* Arabidopsis thaliana   tair10
* Bos taurus     bosTau2
* Bos taurus     bosTau3
* Bos taurus     bosTau4
* Branchiostoma floridae braFlo1
* Caenorhabditis briggsae        CB25
* Caenorhabditis briggsae        cb3
* Caenorhabditis briggsae        cbJul2002
* Caenorhabditis elegans ce10
* Caenorhabditis elegans ce2
* Caenorhabditis elegans ce4
* Caenorhabditis elegans ce6
* Canis lupus familiaris canFam2
* Canis lupus familiaris canFam3
* Ciona intestinalis     CI
* Ciona intestinalis     CI2
* Ciona savignyi CSAV2
* Danio rerio    danRer4
* Danio rerio    danRer5
* Drosophila melanogaster        dm2
* Drosophila melanogaster        dm3
* Drosophila melanogaster        dm6
* Escherichia coli K-12  eschColi
* Gallus gallus  galGal3
* Gallus gallus  galGal4
* Homo sapiens   hg18
* Homo sapiens   hg19
* Homo sapiens   hg38
* Mus musculus   mm10
* Mus musculus   mm8
* Mus musculus   mm9
* Rattus norvegicus      rn4
* Rattus norvegicus      rn5
* Rattus norvegicus      rn6
* Saccharomyces cerevisiae       sacCer1
* Takifugu rubripes      fr1
* Takifugu rubripes      fr2
* Takifugu rubripes      fr3

BED is the simplest file format that can be used for custom annotation with VEP. Annotation files should be stripped of comment lines, sorted in chromosome and position order, compressed using bgzip to be indexed with tabix. The resulting file is then used for annotation.


## Module configuration

The module contains four blocks:

1. Extract information only for specific genome version
2. Create bed file
3. Compress bed file
4. Index compressed bed with tabix

and a config file to specify the genome version of interest, here we picked "hg19"

### Data formats

* Input data
	* Database in tab delimeted text
 
* Output data
	* indexed BED file
	
### Software dependencies
* awk, sed, bgzip, tabix.

### Module parameters
* Replacement and substitution of characters (e.g. spaces, field sepatator) with awk and sed
* Selection of relevant fields:
	- chromosome: genomic region of regulatory element
	- start:	start position
	- end: end position
	- gene_symbol: target gene symbol
	- strand: strand where the variant occurs 
	- type: type of regulatory element (e.g. Transcription factor binding site)
	- experimental_outcome: positive, neutral o negative according to publications
	- database_ID: stable ORegAnno identifier
	- NCBI: PubMed accession for publication supporting variant

## mk-ORegAnno_2\_bed directory structure
````bash
mk-ORegAnno_2_bed/		##Directory for mk module to reformat ORegAnno database
├── bin		##Executables directory.
│   └── create-targets		##Script to print every file required by this module.
├── config.mk		##Configuration file for this module
├── data		##Directory with original database file, input file.
│   └── ORegAnno_2015.tsv		##Database to be reformatted.
├── mkfile		##File in mk format, specifying the rules for building every result requested by bin/create-targets.
└── results		##Storage directory for files built by mkfile.  If it does not exist, it is automatically generated by mkfile.
    ├── ORegAnno_2015.hg19.bed.gz		##output file, sorted and compressed.
    └── ORegAnno_2015.hg19.bed.gz.tbi		##indexed output file.

````

## References


[1]: https://www.ncbi.nlm.nih.gov/pubmed/27268795
1. [McLaren W, Gil L, Hunt SE, et al. The Ensembl Variant Effect Predictor. Genome Biology. 2016;17:122. doi:10.1186/s13059-016-0974-4.][1]

[2]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4702855/
2. [Lesurf R, Cotto KC, Wang G, et al. ORegAnno 3.0: a community-driven resource for curated regulatory annotation. Nucleic Acids Research. 2016;44(Database issue):D126-D132. doi:10.1093/nar/gkv1203][2]

### Author Info
Developed by [Eugenia Zarza](https://www.researchgate.net/profile/Eugenia_Zarza)(eugenia.zarza@gmail.com) for [Winter Genomics](http://www.wintergenomics.com/). 2018.

